src/aws.js

```diff
@@ -71,55 +76,92 @@ api.declare({
   ];
   if (level === 'read-write') {
     objectActions.push(
-      's3:PutObject',
-      's3:DeleteObject'
     );
   }
 
+  var statements = [];
+
+  statements.push({
+    Sid:            'ReadObjectsUnderPrefix',
+    Effect:         'Allow',
+    Action: [
+      's3:GetObject'
+    ],
+    Resource: [
+      'arn:aws:s3:::{{bucket}}/{{prefix}}*'
+        .replace('{{bucket}}', bucket)
+        .replace('{{prefix}}', prefix)
+    ]
+  });
+  statements.push({
+    Sid:            'ListObjectsUnderPrefix',
+    Effect:         'Allow',
+    Action: [
+      's3:ListBucket'
+    ],
+    Resource: [
+      'arn:aws:s3:::{{bucket}}'
+        .replace('{{bucket}}', bucket)
+    ],
+    Condition: {
+      StringLike: {
+        's3:prefix': [
+          '{{prefix}}*'.replace('{{prefix}}', prefix)
+        ]
+      }
+    }
+  });
+  statements.push({
+    Sid:            'GetBucketLocation',
+    Effect:         'Allow',
+    Action: [
+      's3:GetBucketLocation'
+    ],
+    Resource: [
+      'arn:aws:s3:::{{bucket}}'
+        .replace('{{bucket}}', bucket)
+    ]
+  });
+
+  if (level === 'read-write') {
+    statements.push({
+      Sid:            'WriteObjectsUnderPrefix',
+      Effect:         'Allow',
+      Action: [
+        's3:PutObject',
+        's3:DeleteObject'
+      ],
+      Resource: [
+        'arn:aws:s3:::{{bucket}}/{{prefix}}*'
+          .replace('{{bucket}}', bucket)
+          .replace('{{prefix}}', prefix)
+      ]
+    });
+    statements.push({
+      Sid:            'SetAclsForObjectsUnderPrefix',
+      Effect:         'Allow',
+      Action: [
+        's3:PutObjectAcl',
+      ],
+      Resource: [
+        'arn:aws:s3:::{{bucket}}/{{prefix}}*'
+          .replace('{{bucket}}', bucket)
+          .replace('{{prefix}}', prefix)
+      ],
+      Condition: {
+        StringEquals: {
+          's3:x-amz-acl': ['private', 'public-read']
+        }
```

Comment:
I think we should make this restriction on "s3:PutObject" too...
Maybe it should be `['', 'private', 'public-read']` (I'm just guessing that empty string is equal to header missing).

I suspect we'll also want to forbid:
```
x-amz-grant-read	
x-amz-grant-write	
x-amz-grant-read-acp	
x-amz-grant-write-acp	
x-amz-grant-full-control	
```
ie. require the headers are empty...

But I'm not even sure that's enough... because it is possible to specify the ACL in the payload instead of the header...